File: apps/client/src-tauri/crates/command/src/workspaceSkills.rs
#![allow(non_snake_case)]

use std::collections::HashSet;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use thinking_core::{CommandResult, Exception};

const SKILL_FILE: &str = "SKILL.md";
const DESC_MAX: usize = 80;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListSkillsP {
    pub roots: Vec<String>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SkillR {
    pub id: String,
    pub name: String,
    pub description: String,
    pub path: String,
    pub relative: String,
}

fn validate_root(root: &str) -> Option<PathBuf> {
    let trimmed = root.trim();
    if trimmed.is_empty() {
        return None;
    }
    let buf = PathBuf::from(trimmed);
    if buf.is_absolute() && buf.is_dir() {
        Some(buf)
    } else {
        None
    }
}

fn truncate_desc(value: &str) -> String {
    let trimmed = value.split_whitespace().collect::<Vec<_>>().join(" ");
    if trimmed.chars().count() <= DESC_MAX {
        return trimmed;
    }
    let cut: String = trimmed.chars().take(DESC_MAX).collect();
    format!("{cut}…")
}

fn parse_frontmatter(content: &str) -> (Option<String>, Option<String>) {
    let trimmed = content.trim_start();
    if !trimmed.starts_with("---") {
        return (None, None);
    }
    let rest = &trimmed[3..];
    let Some(end) = rest.find("\n---") else {
        return (None, None);
    };
    let block = &rest[..end];
    let mut name = None;
    let mut description = None;
    let mut current_key: Option<&str> = None;
    let mut current_value = String::new();

    fn flush(
        key: Option<&str>,
        value: &str,
        name: &mut Option<String>,
        description: &mut Option<String>,
    ) {
        let Some(key) = key else {
            return;
        };
        let cleaned = value
            .trim()
            .trim_matches(|c| c == '"' || c == '\'')
            .trim()
            .to_string();
        if cleaned.is_empty() {
            return;
        }
        match key {
            "name" => *name = Some(cleaned),
            "description" => *description = Some(cleaned),
            _ => {}
        }
    }

    for line in block.lines() {
        let stripped = line.trim_end();
        if stripped.is_empty() {
            continue;
        }
        if let Some((key, value)) = stripped.split_once(':') {
            let key = key.trim();
            if key == "name" || key == "description" {
                flush(current_key, &current_value, &mut name, &mut description);
                current_key = Some(key);
                let value = value.trim();
                if value == ">" || value == "|-" || value == "|" || value == ">-" {
                    current_value.clear();
                } else {
                    current_value = value.to_string();
                }
                continue;
            }
        }
        if current_key.is_some() {
            if !current_value.is_empty() {
                current_value.push(' ');
            }
            current_value.push_str(stripped.trim());
        }
    }
    flush(current_key, &current_value, &mut name, &mut description);
    (name, description)
}

fn collect_from_skills_dir(root: &Path, skills_dir: &Path, out: &mut Vec<SkillR>, seen: &mut HashSet<String>) {
    let Ok(entries) = std::fs::read_dir(skills_dir) else {
        return;
    };
    for entry in entries.flatten() {
        let path = entry.path();
        if !path.is_dir() {
            continue;
        }
        let skill_path = path.join(SKILL_FILE);
        if !skill_path.is_file() {
            continue;
        }
        let absolute = skill_path.to_string_lossy().to_string();
        if !seen.insert(absolute.clone()) {
            continue;
        }
        let folder_name = path
            .file_name()
            .and_then(|name| name.to_str())
            .unwrap_or("skill")
            .to_string();
        let content = std::fs::read_to_string(&skill_path).unwrap_or_default();
        let (parsed_name, parsed_desc) = parse_frontmatter(&content);
        let relative = skill_path
            .strip_prefix(root)
            .ok()
            .map(|item| item.to_string_lossy().replace('\\', "/"))
            .unwrap_or_else(|| absolute.clone());
        out.push(SkillR {
            id: absolute.clone(),
            name: parsed_name.unwrap_or(folder_name),
            description: truncate_desc(&parsed_desc.unwrap_or_default()),
            path: absolute,
            relative,
        });
    }
}

fn skill_parent_candidates(root: &Path) -> Vec<PathBuf> {
    let mut parents = vec![
        root.join(".agents").join("skills"),
        root.join(".cursor").join("skills"),
        root.join(".qoder").join("skills"),
    ];
    if let Ok(entries) = std::fs::read_dir(root) {
        for entry in entries.flatten().take(40) {
            let path = entry.path();
            if !path.is_dir() {
                continue;
            }
            let name = path
                .file_name()
                .and_then(|item| item.to_str())
                .unwrap_or("");
            if name.starts_with('.') || name == "node_modules" || name == "target" {
                continue;
            }
            parents.push(path.join(".cursor").join("skills"));
            parents.push(path.join("src-tauri").join(".cursor").join("skills"));
            // apps/client/src-tauri/.cursor/skills
            if let Ok(children) = std::fs::read_dir(&path) {
                for child in children.flatten().take(30) {
                    let child_path = child.path();
                    if !child_path.is_dir() {
                        continue;
                    }
                    parents.push(child_path.join(".cursor").join("skills"));
                    parents.push(child_path.join("src-tauri").join(".cursor").join("skills"));
                }
            }
        }
    }
    parents
}

#[tauri::command(rename = "workspaceSkills:list")]
pub async fn workspaceSkillsList(params: ListSkillsP) -> CommandResult<Vec<SkillR>> {
    let roots: Vec<PathBuf> = params
        .roots
        .iter()
        .filter_map(|root| validate_root(root))
        .collect();
    let skills = tokio::task::spawn_blocking(move || {
        let mut out = Vec::new();
        let mut seen = HashSet::new();
        for root in roots {
            for parent in skill_parent_candidates(&root) {
                if parent.is_dir() {
                    collect_from_skills_dir(&root, &parent, &mut out, &mut seen);
                }
            }
        }
        out.sort_by(|a, b| a.name.to_ascii_lowercase().cmp(&b.name.to_ascii_lowercase()));
        out
    })
    .await
    .map_err(|err| Exception::Internal(format!("skills 扫描失败：{err}")))?;
    Ok(skills)
}

File: apps/studio/src/host/capabilities/workspace-skills.ts
import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs'
import path from 'node:path'

/**
 * 工作区技能扫描：对齐 client `workspaceSkills.rs`。
 * 在根下候选目录里找含 `SKILL.md` 的子文件夹，解析 YAML frontmatter。
 */

const SKILL_FILE = 'SKILL.md'
const DESC_MAX = 80
const ROOT_CHILD_CAP = 40
const NESTED_CHILD_CAP = 30

interface WorkspaceSkill {
  id: string
  name: string
  description: string
  path: string
  relative: string
}

function truncateDesc(value: string): string {
  const trimmed = value.split(/\s+/).filter(Boolean).join(' ')
  if ([...trimmed].length <= DESC_MAX) return trimmed
  return `${[...trimmed].slice(0, DESC_MAX).join('')}…`
}

function parseFrontmatter(content: string): { name?: string; description?: string } {
  const trimmed = content.trimStart()
  if (!trimmed.startsWith('---')) return {}

  const rest = trimmed.slice(3)
  const end = rest.indexOf('\n---')
  if (end < 0) return {}

  const block = rest.slice(0, end)
  let name: string | undefined
  let description: string | undefined
  let currentKey: string | undefined
  let currentValue = ''

  function flush() {
    if (!currentKey) return
    const cleaned = currentValue
      .trim()
      .replace(/^['"]|['"]$/g, '')
      .trim()
    if (!cleaned) return
    if (currentKey === 'name') name = cleaned
    if (currentKey === 'description') description = cleaned
  }

  for (const line of block.split(/\r?\n/)) {
    const stripped = line.trimEnd()
    if (!stripped) continue
    const colon = stripped.indexOf(':')
    if (colon > 0) {
      const key = stripped.slice(0, colon).trim()
      if (key === 'name' || key === 'description') {
        flush()
        currentKey = key
        const value = stripped.slice(colon + 1).trim()
        if (value === '>' || value === '|-' || value === '|' || value === '>-') {
          currentValue = ''
        } else {
          currentValue = value
        }
        continue
      }
    }
    if (currentKey) {
      if (currentValue) currentValue += ' '
      currentValue += stripped.trim()
    }
  }
  flush()

  return { name, description }
}

function toPosixRelative(root: string, absolute: string): string {
  const relative = path.relative(root, absolute)
  if (!relative || relative.startsWith('..') || path.isAbsolute(relative)) {
    return absolute.replace(/\\/g, '/')
  }
  return relative.replace(/\\/g, '/')
}

function collectFromSkillsDir(
  root: string,
  skillsDir: string,
  out: WorkspaceSkill[],
  seen: Set<string>
) {
  let entries
  try {
    entries = readdirSync(skillsDir, { withFileTypes: true })
  } catch {
    return
  }

  for (const entry of entries) {
    if (!entry.isDirectory()) continue
    const folder = path.join(skillsDir, entry.name)
    const skillPath = path.join(folder, SKILL_FILE)
    if (!existsSync(skillPath) || !statSync(skillPath).isFile()) continue

    const absolute = skillPath
    if (seen.has(absolute)) continue
    seen.add(absolute)

    let content = ''
    try {
      content = readFileSync(skillPath, 'utf8')
    } catch {
      content = ''
    }
    const parsed = parseFrontmatter(content)
    out.push({
      id: absolute,
      name: parsed.name ?? entry.name,
      description: truncateDesc(parsed.description ?? ''),
      path: absolute,
      relative: toPosixRelative(root, absolute)
    })
  }
}

function skillParentCandidates(root: string): string[] {
  const parents = [
    path.join(root, '.agents', 'skills'),
    path.join(root, '.cursor', 'skills'),
    path.join(root, '.qoder', 'skills')
  ]

  let entries
  try {
    entries = readdirSync(root, { withFileTypes: true }).slice(0, ROOT_CHILD_CAP)
  } catch {
    return parents
  }

  for (const entry of entries) {
    if (!entry.isDirectory()) continue
    const name = entry.name
    if (name.startsWith('.') || name === 'node_modules' || name === 'target') continue

    const child = path.join(root, name)
    parents.push(path.join(child, '.cursor', 'skills'))
    parents.push(path.join(child, 'src-tauri', '.cursor', 'skills'))

    let nested
    try {
      nested = readdirSync(child, { withFileTypes: true }).slice(0, NESTED_CHILD_CAP)
    } catch {
      continue
    }
    for (const nest of nested) {
      if (!nest.isDirectory()) continue
      const nestPath = path.join(child, nest.name)
      parents.push(path.join(nestPath, '.cursor', 'skills'))
      parents.push(path.join(nestPath, 'src-tauri', '.cursor', 'skills'))
    }
  }

  return parents
}

/**
 * 扫描工作区根下的本地技能。
 */
function listSkills(root: string): WorkspaceSkill[] {
  const out: WorkspaceSkill[] = []
  const seen = new Set<string>()
  for (const parent of skillParentCandidates(root)) {
    if (existsSync(parent) && statSync(parent).isDirectory()) {
      collectFromSkillsDir(root, parent, out, seen)
    }
  }
  out.sort(function (a, b) {
    return a.name.localeCompare(b.name, undefined, { sensitivity: 'base' })
  })
  return out
}

export { listSkills, parseFrontmatter }
export type { WorkspaceSkill }
